File: public/codes/data-structure/hash-table.py
# ハッシュテーブル (Hash Table) - Python実装


class HashTable:
    """チェイン法を使用したハッシュテーブル実装"""

    def __init__(self, size=10):
        self.size = size
        self.buckets = [[] for _ in range(size)]  # バケット配列
        self.count = 0  # 格納されている要素数

    def _hash(self, key):
        """ハッシュ関数 - 文字列を配列のインデックスに変換"""
        hash_value = 0
        for char in str(key):
            hash_value += ord(char)
        return hash_value % self.size

    def set(self, key, value):
        """キーと値のペアを挿入 - 平均 O(1)"""
        index = self._hash(key)
        bucket = self.buckets[index]

        # 既存のキーがあるかチェック
        for i, (k, v) in enumerate(bucket):
            if k == key:
                bucket[i] = (key, value)  # 値を更新
                return

        # 新しいキーと値を追加
        bucket.append((key, value))
        self.count += 1

        # 負荷率が高くなったらリサイズ
        if self.count > self.size * 0.75:
            self._resize()

    def get(self, key):
        """キーに対応する値を取得 - 平均 O(1)"""
        index = self._hash(key)
        bucket = self.buckets[index]

        for k, v in bucket:
            if k == key:
                return v

        raise KeyError(f"Key '{key}' not found")

    def delete(self, key):
        """キーと値のペアを削除 - 平均 O(1)"""
        index = self._hash(key)
        bucket = self.buckets[index]

        for i, (k, v) in enumerate(bucket):
            if k == key:
                value = bucket.pop(i)[1]
                self.count -= 1
                return value

        raise KeyError(f"Key '{key}' not found")

    def has(self, key):
        """キーが存在するかチェック - 平均 O(1)"""
        try:
            self.get(key)
            return True
        except KeyError:
            return False

    def keys(self):
        """すべてのキーを取得 - O(n)"""
        keys_list = []
        for bucket in self.buckets:
            for k, v in bucket:
                keys_list.append(k)
        return keys_list

    def values(self):
        """すべての値を取得 - O(n)"""
        values_list = []
        for bucket in self.buckets:
            for k, v in bucket:
                values_list.append(v)
        return values_list

    def items(self):
        """すべてのキーと値のペアを取得 - O(n)"""
        items_list = []
        for bucket in self.buckets:
            for k, v in bucket:
                items_list.append((k, v))
        return items_list

    def get_size(self):
        """ハッシュテーブルのサイズ（要素数）を取得 - O(1)"""
        return self.count

    def is_empty(self):
        """ハッシュテーブルが空かチェック - O(1)"""
        return self.count == 0

    def clear(self):
        """ハッシュテーブルをクリア - O(1)"""
        self.buckets = [[] for _ in range(self.size)]
        self.count = 0

    def _resize(self):
        """ハッシュテーブルのサイズを拡張 - O(n)"""
        old_buckets = self.buckets
        self.size *= 2
        self.buckets = [[] for _ in range(self.size)]
        old_count = self.count
        self.count = 0

        # 既存の要素を再ハッシュ
        for bucket in old_buckets:
            for k, v in bucket:
                self.set(k, v)

    def get_load_factor(self):
        """負荷率を計算 - O(1)"""
        return self.count / self.size

    def show_buckets(self):
        """バケットの状態を表示（デバッグ用）"""
        print("Bucket状態:")
        for i, bucket in enumerate(self.buckets):
            if bucket:
                print(f"Bucket {i}: {bucket}")

    def __str__(self):
        """ハッシュテーブルの内容を表示"""
        pairs = []
        for bucket in self.buckets:
            for k, v in bucket:
                pairs.append(f"'{k}': {v}")
        return "{" + ", ".join(pairs) + "}"

    def __getitem__(self, key):
        """dict風のアクセス"""
        return self.get(key)

    def __setitem__(self, key, value):
        """dict風の代入"""
        self.set(key, value)

    def __delitem__(self, key):
        """dict風の削除"""
        self.delete(key)

    def __contains__(self, key):
        """in演算子のサポート"""
        return self.has(key)

    def __len__(self):
        """len()関数のサポート"""
        return self.count


def count_words(text):
    """単語の出現回数をカウント"""
    word_count = HashTable()
    words = text.lower().split()

    for word in words:
        # 句読点を除去
        word = word.strip('.,!?;:"()[]{}')
        if word:
            current_count = 0
            try:
                current_count = word_count.get(word)
            except KeyError:
                pass
            word_count.set(word, current_count + 1)

    return word_count


class SimpleCache:
    """シンプルなキャッシュシステム（LRU方式）"""

    def __init__(self, max_size=5):
        self.cache = HashTable(max_size)
        self.access_order = []  # LRU用のアクセス順序
        self.max_size = max_size

    def get(self, key):
        """キャッシュから値を取得"""
        try:
            value = self.cache.get(key)
            self._update_access_order(key)
            print(f"キャッシュヒット: {key} = {value}")
            return value
        except KeyError:
            print(f"キャッシュミス: {key}")
            return None

    def set(self, key, value):
        """キャッシュに値を設定"""
        # キャッシュが満杯の場合、最も古いアイテムを削除
        if self.cache.get_size() >= self.max_size and key not in self.cache:
            oldest_key = self.access_order.pop(0)
            self.cache.delete(oldest_key)
            print(f"キャッシュから削除: {oldest_key}")

        self.cache.set(key, value)
        self._update_access_order(key)
        print(f"キャッシュに追加: {key} = {value}")

    def _update_access_order(self, key):
        """アクセス順序を更新"""
        if key in self.access_order:
            self.access_order.remove(key)
        self.access_order.append(key)

    def show_cache(self):
        """現在のキャッシュ状態を表示"""
        print(f"現在のキャッシュ: {self.cache}")
        print(f"アクセス順序: {self.access_order}")


# 使用例
if __name__ == "__main__":
    print("=== 基本的なハッシュテーブル操作 ===")
    hash_table = HashTable(5)

    # 要素の追加
    hash_table.set("name", "田中太郎")
    hash_table.set("age", 25)
    hash_table.set("city", "東京")
    hash_table.set("hobby", "プログラミング")

    print(hash_table)
    print(f"サイズ: {hash_table.get_size()}")

    # 要素の取得
    print(f"名前: {hash_table.get('name')}")
    print(f"年齢: {hash_table.get('age')}")

    # キーの存在チェック
    print(f"emailキーは存在する? {'email' in hash_table}")

    # 要素の削除
    deleted_value = hash_table.delete("hobby")
    print(f"削除された値: {deleted_value}")
    print(hash_table)

    # すべてのキーと値を取得
    print(f"すべてのキー: {hash_table.keys()}")
    print(f"すべての値: {hash_table.values()}")

    # バケットの状態を確認
    hash_table.show_buckets()

    # 応用例1: 単語の出現回数カウント
    print("\n=== 単語の出現回数カウント ===")
    text = "この文章には同じ単語が複数回出現します この文章を解析して単語の出現回数を数えます"
    word_counter = count_words(text)
    print("単語の出現回数:")
    print(word_counter)

    # 応用例2: キャッシュシステム
    print("\n=== シンプルなキャッシュシステム ===")
    cache = SimpleCache(3)
    cache.set("user1", "田中")
    cache.set("user2", "佐藤")
    cache.set("user3", "鈴木")
    cache.show_cache()

    cache.get("user1")  # キャッシュヒット
    cache.set("user4", "高橋")  # user2が削除される
    cache.show_cache()

    # dict風の操作
    print("\n=== dict風の操作 ===")
    ht = HashTable()
    ht["key1"] = "value1"
    ht["key2"] = "value2"
    print(f"ht['key1'] = {ht['key1']}")
    print(f"'key2' in ht: {'key2' in ht}")
    print(f"len(ht): {len(ht)}")
    del ht["key1"]
    print(f"削除後: {ht}")

File: public/codes/data-structure/hash-table.js
// ハッシュテーブル (Hash Table) - JavaScript実装

class HashTable {
  constructor(size = 10) {
    this.size = size;
    this.buckets = new Array(size); // バケット配列
    this.count = 0; // 格納されている要素数

    // 各バケットを配列で初期化（チェイン法用）
    for (let i = 0; i < size; i++) {
      this.buckets[i] = [];
    }
  }

  // ハッシュ関数 - 文字列を配列のインデックスに変換
  _hash(key) {
    let hash = 0;
    for (let i = 0; i < key.length; i++) {
      hash += key.charCodeAt(i);
    }
    return hash % this.size;
  }

  // キーと値のペアを挿入 - 平均 O(1)
  set(key, value) {
    const index = this._hash(key);
    const bucket = this.buckets[index];

    // 既存のキーがあるかチェック
    for (let i = 0; i < bucket.length; i++) {
      if (bucket[i][0] === key) {
        bucket[i][1] = value; // 値を更新
        return;
      }
    }

    // 新しいキーと値を追加
    bucket.push([key, value]);
    this.count++;

    // 負荷率が高くなったらリサイズ
    if (this.count > this.size * 0.75) {
      this._resize();
    }
  }

  // キーに対応する値を取得 - 平均 O(1)
  get(key) {
    const index = this._hash(key);
    const bucket = this.buckets[index];

    for (let i = 0; i < bucket.length; i++) {
      if (bucket[i][0] === key) {
        return bucket[i][1];
      }
    }

    return undefined; // キーが見つからない場合
  }

  // キーと値のペアを削除 - 平均 O(1)
  delete(key) {
    const index = this._hash(key);
    const bucket = this.buckets[index];

    for (let i = 0; i < bucket.length; i++) {
      if (bucket[i][0] === key) {
        const value = bucket[i][1];
        bucket.splice(i, 1);
        this.count--;
        return value;
      }
    }

    return undefined; // キーが見つからない場合
  }

  // キーが存在するかチェック - 平均 O(1)
  has(key) {
    return this.get(key) !== undefined;
  }

  // すべてのキーを取得 - O(n)
  keys() {
    const keys = [];
    for (let bucket of this.buckets) {
      for (let [key, value] of bucket) {
        keys.push(key);
      }
    }
    return keys;
  }

  // すべての値を取得 - O(n)
  values() {
    const values = [];
    for (let bucket of this.buckets) {
      for (let [key, value] of bucket) {
        values.push(value);
      }
    }
    return values;
  }

  // ハッシュテーブルのサイズ（要素数）を取得 - O(1)
  getSize() {
    return this.count;
  }

  // ハッシュテーブルが空かチェック - O(1)
  isEmpty() {
    return this.count === 0;
  }

  // ハッシュテーブルをクリア - O(1)
  clear() {
    for (let i = 0; i < this.size; i++) {
      this.buckets[i] = [];
    }
    this.count = 0;
  }

  // ハッシュテーブルのサイズを拡張 - O(n)
  _resize() {
    const oldBuckets = this.buckets;
    this.size *= 2;
    this.buckets = new Array(this.size);
    this.count = 0;

    // 新しいバケットを初期化
    for (let i = 0; i < this.size; i++) {
      this.buckets[i] = [];
    }

    // 既存の要素を再ハッシュ
    for (let bucket of oldBuckets) {
      for (let [key, value] of bucket) {
        this.set(key, value);
      }
    }
  }

  // ハッシュテーブルの内容を表示 - O(n)
  toString() {
    const pairs = [];
    for (let bucket of this.buckets) {
      for (let [key, value] of bucket) {
        pairs.push(`${key}: ${value}`);
      }
    }
    return `{${pairs.join(", ")}}`;
  }

  // 負荷率を計算 - O(1)
  getLoadFactor() {
    return this.count / this.size;
  }

  // バケットの状態を表示（デバッグ用）
  showBuckets() {
    console.log("Bucket状態:");
    for (let i = 0; i < this.buckets.length; i++) {
      if (this.buckets[i].length > 0) {
        console.log(`Bucket ${i}: ${JSON.stringify(this.buckets[i])}`);
      }
    }
  }
}

// 使用例とハッシュテーブルの応用例

// 基本的な使用例
console.log("=== 基本的なハッシュテーブル操作 ===");
const hashTable = new HashTable(5);

// 要素の追加
hashTable.set("name", "田中太郎");
hashTable.set("age", 25);
hashTable.set("city", "東京");
hashTable.set("hobby", "プログラミング");

console.log(hashTable.toString());
console.log("サイズ:", hashTable.getSize());

// 要素の取得
console.log("名前:", hashTable.get("name"));
console.log("年齢:", hashTable.get("age"));

// キーの存在チェック
console.log("emailキーは存在する?", hashTable.has("email"));

// 要素の削除
console.log("削除された値:", hashTable.delete("hobby"));
console.log(hashTable.toString());

// すべてのキーと値を取得
console.log("すべてのキー:", hashTable.keys());
console.log("すべての値:", hashTable.values());

// バケットの状態を確認
hashTable.showBuckets();

// 応用例1: 単語の出現回数をカウント
console.log("\n=== 単語の出現回数カウント ===");
function countWords(text) {
  const wordCount = new HashTable();
  const words = text.toLowerCase().split(/\s+/);

  for (let word of words) {
    const count = wordCount.get(word) || 0;
    wordCount.set(word, count + 1);
  }

  return wordCount;
}

const text =
  "この文章には同じ単語が複数回出現します この文章を解析して単語の出現回数を数えます";
const wordCounter = countWords(text);
console.log("単語の出現回数:");
console.log(wordCounter.toString());

// 応用例2: キャッシュシステム
console.log("\n=== シンプルなキャッシュシステム ===");
class SimpleCache {
  constructor(maxSize = 5) {
    this.cache = new HashTable(maxSize);
    this.accessOrder = []; // LRU用のアクセス順序
    this.maxSize = maxSize;
  }

  get(key) {
    const value = this.cache.get(key);
    if (value !== undefined) {
      // アクセス順序を更新
      this._updateAccessOrder(key);
      console.log(`キャッシュヒット: ${key} = ${value}`);
      return value;
    }
    console.log(`キャッシュミス: ${key}`);
    return null;
  }

  set(key, value) {
    // キャッシュが満杯の場合、最も古いアイテムを削除
    if (this.cache.getSize() >= this.maxSize && !this.cache.has(key)) {
      const oldestKey = this.accessOrder.shift();
      this.cache.delete(oldestKey);
      console.log(`キャッシュから削除: ${oldestKey}`);
    }

    this.cache.set(key, value);
    this._updateAccessOrder(key);
    console.log(`キャッシュに追加: ${key} = ${value}`);
  }

  _updateAccessOrder(key) {
    // キーを順序リストから削除してから末尾に追加
    const index = this.accessOrder.indexOf(key);
    if (index > -1) {
      this.accessOrder.splice(index, 1);
    }
    this.accessOrder.push(key);
  }

  showCache() {
    console.log("現在のキャッシュ:", this.cache.toString());
    console.log("アクセス順序:", this.accessOrder);
  }
}

const cache = new SimpleCache(3);
cache.set("user1", "田中");
cache.set("user2", "佐藤");
cache.set("user3", "鈴木");
cache.showCache();

cache.get("user1"); // キャッシュヒット
cache.set("user4", "高橋"); // user2が削除される
cache.showCache();
